### ISSUE TITLE
hidbootloader crashes after inserting NISSE in bootloader mode

### ISSUE THREAD

**Issue by listx** [esrille/new-keyboard#7]
Steps to reproduce (on Mac OSX Yosemite):

- Build HIDBootloader with Qt Creator 5.
- Connect NISSE via USB while holding ESC key (red LED blinking).
- Run HIDBootloader (Command + R).

Output:

```
Starting /Users/larver/nisse/new-keyboard/firmware/third_party/mla_v2013_12_20/apps/usb/device/bootloaders/utilities/build-HIDBootloader-Desktop_Qt_5_5_1_clang_64bit-Debug/Bootloader/mac/HIDBootloader.app/Contents/MacOS/HIDBootloader...
2015-10-31 23:44:39.187 HIDBootloader[735:5914] An uncaught exception was raised
2015-10-31 23:44:39.187 HIDBootloader[735:5914] -[NSTaggedPointerString getCharacters:range:]: Range {0, 32} out of bounds; string length 3
2015-10-31 23:44:39.187 HIDBootloader[735:5914] (
	0   CoreFoundation                      0x00007fff921b303c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff8937376e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff921b2eed +[NSException raise:format:] + 205
	3   CoreFoundation                      0x00007fff9215ab97 -[NSTaggedPointerString getCharacters:range:] + 263
	4   CoreFoundation                      0x00007fff920875b6 __CFStringEncodeByteStream + 3638
	5   CoreFoundation                      0x00007fff92062a6b CFStringGetBytes + 795
	6   HIDBootloader                       0x000000010d141620 get_string_property_utf8 + 160
	7   HIDBootloader                       0x000000010d13fcd9 make_path + 73
	8   HIDBootloader                       0x000000010d13f9fb hid_enumerate + 587
	9   HIDBootloader                       0x000000010d13b5c5 _ZN4Comm7PollUSBEv + 37
	10  HIDBootloader                       0x000000010d12c104 _ZN10MainWindowC2EP7QWidget + 3444
	11  HIDBootloader                       0x000000010d12d7fd _ZN10MainWindowC1EP7QWidget + 29
	12  HIDBootloader                       0x000000010d13aa0f main + 223
	13  HIDBootloader                       0x000000010d129984 start + 52
)
2015-10-31 23:44:39.188 HIDBootloader[735:5914] *** Terminating app due to uncaught exception 'NSRangeException', reason: '-[NSTaggedPointerString getCharacters:range:]: Range {0, 32} out of bounds; string length 3'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff921b303c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff8937376e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff921b2eed +[NSException raise:format:] + 205
	3   CoreFoundation                      0x00007fff9215ab97 -[NSTaggedPointerString getCharacters:range:] + 263
	4   CoreFoundation                      0x00007fff920875b6 __CFStringEncodeByteStream + 3638
	5   CoreFoundation                      0x00007fff92062a6b CFStringGetBytes + 795
	6   HIDBootloader                       0x000000010d141620 get_string_property_utf8 + 160
	7   HIDBootloader                       0x000000010d13fcd9 make_path + 73
	8   HIDBootloader                       0x000000010d13f9fb hid_enumerate + 587
	9   HIDBootloader                       0x000000010d13b5c5 _ZN4Comm7PollUSBEv + 37
	10  HIDBootloader                       0x000000010d12c104 _ZN10MainWindowC2EP7QWidget + 3444
	11  HIDBootloader                       0x000000010d12d7fd _ZN10MainWindowC1EP7QWidget + 29
	12  HIDBootloader                       0x000000010d13aa0f main + 223
	13  HIDBootloader                       0x000000010d129984 start + 52
)
libc++abi.dylib: terminating with uncaught exception of type NSException
The program has unexpectedly finished.
/Users/larver/nisse/new-keyboard/firmware/third_party/mla_v2013_12_20/apps/usb/device/bootloaders/utilities/build-HIDBootloader-Desktop_Qt_5_5_1_clang_64bit-Debug/Bootloader/mac/HIDBootloader.app/Contents/MacOS/HIDBootloader crashed
```

My NISSE's FN-F1 output is:

```
esrille nisse
rev. 5
ver. 0.17
copyright 2013-2015 esrille inc.
f2 pc
f3 us
f4 roma
f5 d0
f6 c
f7 ms
f8 c-c
f9 off
```

**Comment by ShikiOkasaka** [comment]
I think this a known issue of the HIDBootloader in mla_v2013_12_20.
Please run HIDBootloader *before* connecting NISSE while holding ESC key.

**Comment by listx** [comment]
My google searches led me to http://hieroglyphe.blog47.fc2.com/blog-entry-99.html and although I cannot read Japanese I can see that the person there did the reverse (start HIDBootloader first, and then connect with ESC key held down). But, HIDBootloader still says "Device not detected" (same message as when it starts up).

**Comment by listx** [comment]
I will clarify: HIDBootloader runs fine on its own (no NISSE attached), but says "Device not detected". Connecting NISSE in bootloader mode (ESC key) has no effect (HIDBootloader can't see the NISSE).

**Comment by ShikiOkasaka** [comment]
Thanks for the clarification. The MacBook I have here is a bit old and I cannot reproduce the issue. I will get a newer Mac and investigate the issue as soon as I can, and let you know what I find out. 

Several ideas occurred to me are:
* Reboot the Mac to see HIDBootloader might work,
* Use the newer HIDBootloader in the Microchip Libraries for Applications:
http://www.microchip.com/pagehandler/en-us/devtools/mla/home.html

If you have a Windows or Linux machine and if possible, please try and see if HIDBootloader works with your NISSE.

Thank you,

**Comment by listx** [comment]
I did the 2nd option and downloaded mla_v2015_08_10_osx_installer.dmg (newest one) and it worked! Here are the exact steps I used below:

1. I built the newer HIDBootloader binary version 2.13.
2. I ran the HIDBootloader with Command+R.
3. I connected the keyboard by holding down ESC key, and this recognized the keyboard (HIDBootloader was responsive!).
4. It was then a matter of following the reset of the firmware update steps on the Nisse support page here https://www.esrille.com/keyboard/support.html#firmware (import the .hex file I generated from MPLAB IDE with `Program -> Erase/Program/Verify Device` and then press `Program -> Reset Device`).

Thanks for your help. Closing!
